### DISCUSSION TITLE
Sofar - Further Development

### DISCUSSION THREAD

**Discussion by wills106** [wills106/homeassistant-solax-modbus#655] [Ideas]
When using Time of Use settings on the Inverter what are you able to set?

TOU_ID
TOU_On_Off_Control
TOU_Charge_Start
TOU_Charge_End
TOU_Charge_Target_SOC
TOU_Charge_Power
TOU_Executed_Date_Start
TOU_Executed_Date_End

What can you set TOU_Charge_Power to? Min and Max

Edit:
How many TOU ID's do you have? Is it 4 like Timing ID?

Edit 2:
What options do you have under "TOU_On_Off_Control" Just Enabled / Disabled?

**Comment by josefkrtil** [comment]
informations from manual:
![image](https://github.com/wills106/homeassistant-solax-modbus/assets/34856560/0ae3bd31-7768-42c0-b55f-f1b05de1d3f0)
Charge power is power what is use for charging.

**Comment by dusankonder** [comment]
Hi, 
i have add Inverter Fault sate to my sofar configuration. when you wont to add.
This is the code: plugin_sofar.py
SofarModbusSensorEntityDescription(
        name = "Fault 1",
        key = "Fault 1",
        register = 0x405,
        newblock = True,
        scale = { 0: "No error",
                  1: "ID01 Grid Over Voltage Protection",
                  2: "ID02 Grid Under Voltage Protection",
                  4: "ID03 Grid Over Frequency Protection",
                  8: "ID04 Grid Under Frequency Protection",
                  16: "ID05 Leakage current fault",
                  32: "ID06 High penetration error",
                  64: "ID07 Low penetration error",
		  128: "ID08 Islanding error",
		  256: "ID09 Grid voltage transient value overvoltage 1",
		  512: "ID10 Grid voltage transient value overvoltage 2",
		  1024: "ID11 Grid line voltage error",
		  2048: "ID12 Inverter voltage error",
		  4096: "ID13 Anti-backflow overload",
		  8192: "ID14",
		  16384: "ID15",
		  32768: "ID16", },
        allowedtypes = HYBRID | PV,
		entity_category = EntityCategory.DIAGNOSTIC,
    ),
	SofarModbusSensorEntityDescription(
        name = "Fault 2",
        key = "Fault 2",
        register = 0x406,
        newblock = True,
        scale = { 0: "No error",
                  1: "ID17 Grid current sampling error",
                  2: "ID18 Grid current DC component sampling error (AC side)",
                  4: "ID19 Grid voltage sampling error (DC side)",
                  8: "ID20 Grid voltage sampling error (AC side)",
                  16: "ID21 Leakage current sampling error (DC side)",
                  32: "ID22 Leakage current sampling error (AC side)",
                  64: "ID23 Load voltage DC component sampling error",
		  128: "ID24 DC input current sampling error",
		  256: "ID25 DC component sampling error of grid current (DC side)",
		  512: "ID26 DC input branch current sampling error",
		  1024: "ID27",
		  2048: "ID28",
		  4096: "ID29 Leakage current consistency error",
		  8192: "ID30 Grid voltage consistency error",
		  16384: "ID31 DCI consistency error",
		  32768: "ID32", },
        allowedtypes = HYBRID | PV,
		entity_category = EntityCategory.DIAGNOSTIC,
    ),
	SofarModbusSensorEntityDescription(
        name = "Fault 3",
        key = "Fault 3",
        register = 0x407,
        newblock = True,
        scale = { 0: "No error",
                  1: "ID033 SPI communication error (DC side)",
                  2: "ID034 SPI communication error (AC side)",
                  4: "ID035 Chip error (DC side)",
                  8: "ID036 Chip error (AC side)",
                  16: "ID037 Auxiliary power error",
                  32: "ID038 Inverter soft start failure",
                  64: "ID039",
		  128: "ID040",
		  256: "ID041 Relay detection failure",
		  512: "ID042 Low insulation impedance",
		  1024: "ID043 Grounding error",
		  2048: "ID044 Input mode setting error",
		  4096: "ID044 ID045 CT error",
		  8192: "ID046 Input reversal error",
		  16384: "ID047 Parallel error",
				  32768: "ID048 Serial number error", },
        allowedtypes = HYBRID | PV,
		entity_category = EntityCategory.DIAGNOSTIC,
    ),
	SofarModbusSensorEntityDescription(
        name = "Fault 4",
        key = "Fault 4",
        register = 0x408,
        newblock = True,
        scale = { 0: "No error",
                  1: "ID049 Battery temperature protection",
                  2: "ID050 Heat sink 1 temperature protection",
                  4: "ID051 Heater 2 temperature protection",
                  8: "ID052 Heater 3 temperature protection",
                  16: "ID053 Heatsink 4 temperature protection",
                  32: "ID054 Heatsink 5 temperature protection",
                  64: "ID055 Radiator 6 temperature protection",
		  128: "ID056",
		  256: "ID057 Ambient temperature 1 protection",
		  512: "ID058 Ambient temperature 2 protection",
		  1024: "ID059 Module 1 temperature protection",
		  2048: "ID060 Module 2 temperature protection",
		  4096: "ID061 Module 3 temperature protection",
		  8192: "ID062 Module temperature difference is too large",
		  16384: "ID063",
		  32768: "ID064", },
        allowedtypes = HYBRID | PV,
		entity_category = EntityCategory.DIAGNOSTIC,
    ),
	SofarModbusSensorEntityDescription(
        name = "Fault 5",
        key = "Fault 5",
        register = 0x409,
        newblock = True,
        scale = { 0: "No error",
                  1: "ID065 Bus voltage RMS unbalance",
                  2: "ID066 Bus voltage transient value unbalance",
                  4: "ID067 Undervoltage of busbar during grid connection",
                  8: "ID068 Bus bar low voltage",
                  16: "ID069 PV overvoltage",
                  32: "ID070 Battery over-voltage",
                  64: "ID071 LLCBus overvoltage protection",
		  128: "ID072 Inverter bus voltage RMS software overvoltage",
		  256: "ID073 Inverter bus voltage transient value software overvoltage",
		  512: "ID074 Flying Cross Capacitor Overvoltage Protection",
		  1024: "ID075 Flying Cross capacitor undervoltage protection",
		  2048: "ID076",
		  4096: "ID077",
		  8192: "ID078",
		  16384: "ID079",
		  32768: "ID080", },
        allowedtypes = HYBRID | PV,
		entity_category = EntityCategory.DIAGNOSTIC,
    ),
	SofarModbusSensorEntityDescription(
        name = "Fault 6",
        key = "Fault 6",
        register = 0x040A,
        newblock = True,
        scale = { 0: "No error",
                  1: "ID081 Battery overcurrent software protection",
                  2: "ID082 Dci overcurrent protection",
                  4: "ID083 Output transient current protection",
                  8: "ID084 BuckBoost software overcurrent",
                  16: "ID085 Output RMS current protection",
                  32: "ID086 PV instantaneous current overcurrent software protection",
                  64: "ID087 PV parallel uneven current",
		  128: "ID088 Output current unbalance",
		  256: "ID089 PV software overcurrent protection",
		  512: "ID090 Balanced circuit overcurrent protection",
		  1024: "ID091 Resonance protection",
		  2048: "ID092",
		  4096: "ID093",
		  8192: "ID094",
		  16384: "ID095",
		  32768: "ID096", },
        allowedtypes = HYBRID | PV,
		entity_category = EntityCategory.DIAGNOSTIC,
    ),
	SofarModbusSensorEntityDescription(
        name = "Fault 7",
        key = "Fault 7",
        register = 0x040B,
        newblock = True,
        scale = { 0: "No error",
                  1: "ID097 LLC bus hardware overvoltage",
                  2: "ID098 Inverter bus hardware overvoltage",
                  4: "ID099 BuckBoost hardware overcurrent",
                  8: "ID100 Battery hardware overcurrent",
                  16: "ID101",
                  32: "ID102 PV hardware overcurrent",
                  64: "ID103 AC output hardware overcurrent",
		  128: "ID104",
		  256: "ID105 Power meter error",
		  512: "ID106 Serial number model error",
		  1024: "ID107",
		  2048: "ID108",
		  4096: "ID109",
		  8192: "ID110 Overload protection 1",
		  16384: "ID111 Overload protection 2",
		  32768: "ID112 Overload protection 3", },
        allowedtypes = HYBRID | PV,
		entity_category = EntityCategory.DIAGNOSTIC,
    ),
	SofarModbusSensorEntityDescription(
        name = "Fault 8",
        key = "Fault 8",
        register = 0x040C,
        newblock = True,
        scale = { 0: "No error",
                  1: "ID113 Overtemperature derating",
                  2: "ID114 Frequency down load",
                  4: "ID115 Frequency loading",
                  8: "ID116 Voltage down load",
                  16: "ID117 Voltage loading",
                  32: "ID118",
                  64: "ID119",
		  128: "ID120",
		  256: "ID121 Lightning protection failure (DC)",
		  512: "ID122 Lightning protection failure (AC)",
		  1024: "ID123",
		  2048: "ID124 Battery low voltage protection",
		  4096: "ID125 Battery low voltage shutdown",
		  8192: "ID126 Battery low voltage pre-alarm",
		  16384: "ID127",
		  32768: "ID128", },
        allowedtypes = HYBRID | PV,
		entity_category = EntityCategory.DIAGNOSTIC,
    ),
	SofarModbusSensorEntityDescription(
        name = "Fault 9",
        key = "Fault 9",
        register = 0x040D,
        newblock = True,
        scale = { 0: "No error",
                  1: "ID129 Output hardware overcurrent permanent fault",
                  2: "ID130 Bus overvoltage permanent fault",
                  4: "ID131 Bus hardware over-voltage permanent fault",
                  8: "ID132 PV uneven flow permanent fault",
                  16: "ID133 Battery overcurrent permanent fault in EPS mode",
                  32: "ID134 Output transient overcurrent permanent fault",
                  64: "ID135 Output current unbalance permanent fault",
		  128: "ID136",
		  256: "ID137 Input mode setting error permanent fault",
		  512: "ID138 Input overcurrent permanent fault",
		  1024: "ID139 Input hardware overcurrent permanent fault",
		  2048: "ID140 Relay permanent fault",
		  4096: "ID141 Bus unbalance permanent fault",
		  8192: "ID142 Lightning protection permanent fault - DC side",
		  16384: "ID143 Lightning protection permanent fault - AC side",
		  32768: "ID144", },
        allowedtypes = HYBRID | PV,
		entity_category = EntityCategory.DIAGNOSTIC,
    ),
	SofarModbusSensorEntityDescription(
        name = "Fault 10",
        key = "Fault 10",
        register = 0x040E,
        newblock = True,
        scale = { 0: "No error",
                  1: "ID145 USB fault",
                  2: "ID146 WIFI fault",
                  4: "ID147 Bluetooth fault",
                  8: "ID148 RTC clock fault",
                  16: "ID149 Communication board EEPROM error",
                  32: "ID150 Communication board FLASH error",
                  64: "ID151",
		  128: "ID152 Safety regulation version error",
		  256: "ID153 SCI communication error (DC side)",
		  512: "ID154 SCI communication error (AC side)",
		  1024: "ID155 SCI communication error (convergence board side)",
		  2048: "ID156 Software version inconsistency",
		  4096: "ID157 Lithium battery 1 communication error",
		  8192: "ID158 Li-ion battery 2 communication error",
		  16384: "ID159 Lithium battery 3 communication error",
		  32768: "ID160 Lithium battery 4 communication failure", },
        allowedtypes = HYBRID | PV,
		entity_category = EntityCategory.DIAGNOSTIC,
    ),
	SofarModbusSensorEntityDescription(
        name = "Fault 11",
        key = "Fault 11",
        register = 0x040F,
        newblock = True,
        scale = { 0: "No error",
                  1: "ID161 Forced shutdown",
                  2: "ID162 Remote shutdown",
                  4: "ID163 Drms0 shutdown",
                  8: "ID164",
                  16: "ID165 Remote down load",
                  32: "ID166 Logic interface down load",
                  64: "ID167 Anti-Reverse Flow Downgrade",
		  128: "ID168",
		  256: "ID169 Fan 1 failure",
		  512: "ID170 Fan 2 failure",
		  1024: "ID171 Fan 3 failure",
		  2048: "ID172 Fan 4 failure",
		  4096: "ID173 Fan 5 failure",
		  8192: "ID174 Fan 6 failure",
		  16384: "ID175 Fan 7 fault",
		  32768: "ID176 Meter communication failure", },
        allowedtypes = HYBRID | PV,
		entity_category = EntityCategory.DIAGNOSTIC,
    ),
	SofarModbusSensorEntityDescription(
        name = "Fault 12",
        key = "Fault 12",
        register = 0x0410,
        newblock = True,
        scale = { 0: "No error",
                  1: "ID177 BMS over-voltage alarm",
                  2: "ID178 BMS undervoltage alarm",
                  4: "ID179 BMS high temperature alarm",
                  8: "ID180 BMS low temperature alarm",
                  16: "ID181 BMS charge/discharge overload alarm",
                  32: "ID182 BMS short circuit alarm",
                  64: "ID183 BMS version inconsistency",
		  128: "ID184 BMS CAN version inconsistency",
		  256: "ID185 BMS CAN version is too low",
		  512: "ID186",
		  1024: "ID187",
		  2048: "ID188",
		  4096: "ID189 Arc device communication failure",
		  8192: "ID190 DC arc alarm fault",
		  16384: "ID191 PID repair failed",
		  32768: "ID192 PLC module heartbeat loss", },
        allowedtypes = HYBRID | PV,
		entity_category = EntityCategory.DIAGNOSTIC,
    ),
